### PR TITLE
Implement conversion for stablehlo.select and add Where Op

### LINKED ISSUE #639: [Ops] Add TTIR where OP
We should add `where` OP to TTIR and TTNN dialects and then map it to [ttnn.where](https://docs.tenstorrent.com/ttnn/latest/ttnn/ttnn/where.html) OP.

### LINKED ISSUE #640: [MNIST] Implement conversion of stablehlo.select OP to TTIR dialect
`stablehlo.select` OP specification can be found [here](https://openxla.org/stablehlo/spec#select).

Requires #639

### PR DESCRIPTION
**Pull request by uazizTT**
Fixes #639 
Fixes #640

### CODE REVIEW

**Review by AleksKnezevic — APPROVED** [review]

**Comment by nvukobratTT** [comment]
@dgolubovicTT seems like a good op we should map from FFE as well :))

File: lib/Target/TTNN/TTNNToFlatbuffer.cpp
Comment (mmanzoorTT) [inline lib/Target/TTNN/TTNNToFlatbuffer.cpp:342]: Please add a case for `WhereOp` in `emitTTNNOperation` function.

File: runtime/lib/ttnn/operations/eltwise/where/where.cpp
Comment (mmanzoorTT) [inline runtime/lib/ttnn/operations/eltwise/where/where.cpp:10]: Please change the name to `runWhereOp`

File: lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp
Comment (AleksKnezevic) [reply lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp:422]: I think so, it's still a 1:1 mapping, just with a different name. What I don't think we should be doing in the translation layer is decomposing into multiple ops (i.e. `select -> where if dim==0 else select -> some_other_op`) or decomposing into multiple ops (i.e. `select -> transpose + where` but curious to hear others' opinion.

Comment (nvukobratTT) [reply lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp:422]: > I think so, it's still a 1:1 mapping, just with a different name. What I don't think we should be doing in the translation layer is decomposing into multiple ops (i.e. `select -> where if dim==0 else select -> some_other_op`) or decomposing into multiple ops (i.e. `select -> transpose + where` but curious to hear others' opinion.

This sounds good, but I'm not sure if we can manage to do so for every case in indexing op types.. This will be mostly driven by design from TTNN. If they manage to do 1:1 mapping with PyTorch, we'll probably be able to do the same. However, I'm not certain if this is going to happen sooner or later.. 

In any case, we should force clean mapping from the start. And depending on TTNN achievements, we should probably discuss further strategies for all frontends :))

Comment (svuckovicTT) [reply lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp:422]: Sounds good, thanks guys!

Comment (uazizTT) [reply lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp:422]: We have come across a couple of cases where a 1-1 mapping is not possible, so for that we decided to create custom passes that runs after the StableHLOToTTIR conversion but before the TTIRToTTNN conversion. 

- First case is pattern matching for converting 3 ops (constantOp, broadcastOp and maximumOp) to a single ReluOp. 

- Second case Defne is working on where the stablehlo.gatherOp will first be converted to ttir.gatherOp and then converted to ttir.embeddingsOP in a custom pass. 

As this list grows and we have multiple of these custom passes, maybe should convert them all into a single pass, so it becomes a two step conversion:

```
FrontEnd Dialect -> TTIR 1-1 Conversion
TTIR 1-1 Conversion -> TTIR Native Dialect (Legalization for all Thin Conversions) 
TTIR Native Dialect -> TTIRToTTNN Dialect
```

Comment (nvukobratTT) [reply lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp:422]: @uazizTT I think most of these cases will be covered with decomposition and fusing passes on the TTIR level, just to reflect all backends, not only stable hlo. But I think this is still TBD/WIP :))

Comment (nsmithtt) [reply lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp:422]: > * First case is pattern matching for converting 3 ops (`constantOp`, `broadcastOp` and `maximumOp`) to a single `ReluOp`.

I think this pattern match should have probably been done in TTIR.  We can port it.

> * Second case Defne is working on where the `stablehlo.gatherOp` will first be converted to `ttir.gatherOp` and then converted to `ttir.embeddingsOP` in a custom pass.

This one sounds OK

Comment (sdjordjevicTT) [reply lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp:698]: @jserbedzijaTT is adding a decompose pass in TTIR:
https://github.com/tenstorrent/tt-mlir/pull/969

We should utilize that single place for decompositions.

Similarly, we need to come up with the fusing pass as well:
https://github.com/tenstorrent/tt-mlir/issues/1059

@uazizTT can you please help me understand why we need this specific fusion right now? I am trying to understand the priority for its need right now, so we can come up with the proper solution for it.

Comment (mrakitaTT) [reply lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp:698]: I've just now caught up with this whole conversation, let me try to summarize the key points here because I would like to share this with the whole third party crew so we are all in sync. **Please everyone let me know if you would add or change something.**
- TTIR dialect should be general enough to cover 1:1 mappings from all third party dialects.
- In case when the same op is differently named between dialects, TTIR dialect op name should be chosen as the most common name used in popular third party dialects, to facilitate adoption. For example if all third party dialects use `select`, we shouldn't use `where` only because torch and metal ttnn use `where`. TTNN dialect names on the other hand should match the metal ttnn ops names.
- Situations where multiple simpler ops can be fused into one complex op should be handled like this:
  - We should add the complex op to TTIR dialect too, so it is general enough to support direct lowering from another dialect which has such op.
  - We should not implement fusion directly during conversion from specific third party dialect to TTTIR, because then we would have to write such fusion for each third party dialect. Instead, we should implement it inside fusion pass in TTIR (#1059).
- In situations where third party dialect supports some complex op which is not supported by metal ttnn (and they don't plan to support it), we need to decompose it into set of supported ttnn ops (if possible). In that case we still want to create the said complex op in TTIR dialect and to convert from third party dialect into TTIR 1:1, for all reasons about generality stated above, but then the decomposition should be implemented inside decomposition pass in TTIR (#969).
  - One exception here is when op is supported by metal ttnn but it doesn't currently work or is lacking some functionality, and we can decompose it into set of different ops which are working fine. That should not be implemented inside decomposition pass! Instead, that should be implemented inside different pass used for workarounds (this pass is currently WIP). Before doing any workarounds, the procedure for Metal issues should be followed first. (I'll follow up with that separately and link it here).
- This should cover most cases, any other situation that arises should be reviewed on case by case basis so we can refine these guidelines.

Unresolving thread for more visibility.

Comment (nvukobratTT) [reply lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp:698]: > In case when the same op is differently named between dialects, TTIR dialect op name should be chosen as the most common name used in popular third party dialects, to facilitate adoption. For example if all third party dialects use select, we shouldn't use where only because torch and metal ttnn use where. TTNN dialect names on the other hand should match the metal ttnn ops names.

I partially disagree with this. From my perspective, TTIR should be as wide as possible, and users from any framework should use a set of ops that can be directly mapped from refferent framework. From MLIR end, we should map that kind of ops to the common set we further use (e.g. wide vs narrow TTIR).

Comment (mrakitaTT) [reply lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp:698]: How is the part you've quoted in clash with that? You think we should have both select and where op in TTIR? That seems too excessive to me tbh, TTIR would just become brainless mashup of all third party dialects jumbled together. No other dialect has duplicated ops, duplicated ops = duplicated code and it introduces confusion to the user which op should be used.

File: runtime/lib/ttnn/operations/eltwise/where/where.cpp
Comment (mrakitaTT) [inline runtime/lib/ttnn/operations/eltwise/where/where.cpp:1]: Runtime follows op organization from the metal codebase and `where` op is part of pointwise ternary ops group (https://github.com/tenstorrent/tt-metal/tree/main/ttnn/cpp/ttnn/operations/eltwise/ternary).

Can you please rename the folder from `where` to `ternary` (`runtime/lib/ttnn/operations/eltwise/ternary`)?  I am okay with having separate `where.h` and `where.cpp` from `ternary.h` and `ternary.cpp` because that is what Metal currently does too.

FYI @jnie-TT

File: runtime/lib/ttnn/operations/eltwise/where/where.cpp
Comment (mrakitaTT) [inline runtime/lib/ttnn/operations/eltwise/where/where.cpp:12]: Since `where` op is treated uniquely from other ternary ops, there is no need to use lambda here, just move those arguments from lambda to the `runWhereOp` function directly. In case in the future if we want to invoke `ttnn:where` with a different set of arguments, we will just have a `runWhereOp` with a different signature, similar to how Metal does with different `invoke` functions (https://github.com/tenstorrent/tt-metal/blob/main/ttnn/cpp/ttnn/operations/eltwise/ternary/where.cpp).

File: runtime/lib/ttnn/operations/eltwise/where/where.h
Comment (mrakitaTT) [inline runtime/lib/ttnn/operations/eltwise/where/where.h:5]: Include guards should be named based on the full path in a project's source tree to guarantee uniqueness, I think we even had a clang rule for this. See more here: https://google.github.io/styleguide/cppguide.html#The__define_Guard

I see that other headers in `runtime/lib/ttnn/operations` break this rule too. FYI @jnie-TT

File: test/ttmlir/Conversion/StableHLOToTTIR/select_op.mlir
Comment (mrakitaTT) [inline test/ttmlir/Conversion/StableHLOToTTIR/select_op.mlir:9]: Can you please improve these checks in tests that you've added? See this message: https://tenstorrent.slack.com/archives/C07A30HRX7C/p1727698795548319

**Review by mrakitaTT — CHANGES_REQUESTED** [review]

File: lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp
Comment (nvukobratTT) [reply lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp:698]: Personal preference, but it makes more sense to me to have both Select and Where ops. 

This way, folks coming from different frameworks will have an easier time adjusting to TTIR ops. In the background, we can do aliases and actually fallback to use only a single op. E.g. from TTIR wide (Select, Where, etc.) to TTIR narrow (Select). 

I don't see how and why this introduces confusion and brainless mashup for all third parties. As mentioned, it enables our users to to have easier time fitting in into TTIR.

Comment (mrakitaTT) [reply lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp:698]: > This way, folks coming from different frameworks will have an easier time adjusting to TTIR ops.

Not necessarily, you are trying to save them time finding out which op in our dialect to use, but on the other hand you will lose them time trying to figure out why do we have both select and where op and which one should they use. Imagine if every third party dialect used different name, you would have ten functionally same ops in TTIR with different names. So from the user side I don't think you are doing them any good, and then on our side you are making our code more messy. Every dialect has specs and since we are providing 1:1 mapping for each op it should be fairly easy for them to figure out which op to use.

**Comment by kmitrovicTT** [comment]
@uazizTT does this solve https://github.com/tenstorrent/tt-mlir/issues/1009? Is so, please link it. Also, [excel](https://docs.google.com/spreadsheets/d/15JbTmYDccsJgeD_mJa-V6vvyD3e6IGCKZvJhLpkZr5I/edit?gid=0#gid=0) row for `select` is marked blue, no PR and assignee set to it.

File: lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp
Comment (sdjordjevicTT) [reply lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp:698]: @mrakitaTT you nicely summarized everything. It would be ideal if we could capture your summary somewhere, though I’m not certain if our public docs are the right spot for it.

On the topic of having one or multiple TTIR ops for the same function, this seems to come down to individual preference. Perhaps there’s a middle ground we could explore. For example, Torch defines both clamp and clip operations, which perform the same function, with clip simply acting as an alias for clamp:

	•	[torch.clamp](https://pytorch.org/docs/stable/generated/torch.clamp.html)
	•	[torch.clip](https://pytorch.org/docs/stable/generated/torch.clip.html)

This approach could offer flexibility without adding redundancy.

**Comment by svuckovicTT** [comment]
@kmitrovicTT I wasn't aware there was another place where we tracked ops... @sdjordjevicTT @nvukobratTT can you folks pick one to be the source of truth (either github issues or the spreadsheet) to avoid issues of multiple people implementing the same things?

**Comment by nvukobratTT** [comment]
> @kmitrovicTT I wasn't aware there was another place where we tracked ops... @sdjordjevicTT @nvukobratTT can you folks pick one to be the source of truth (either github issues or the spreadsheet) to avoid issues of multiple people implementing the same things?

From my side, GitHub should be the main source of truth. 

It's available for everyone, and it's regularly updated. Other docs, sheets, etc. should be on a team/personal level, and shouldn't be interpreted for the latest state of op support.

Other thoughts @sdjordjevicTT @AleksKnezevic?

**Comment by mrakitaTT** [comment]
@svuckovicTT @nvukobratTT There was a parallel discussion about this today and we've agreed with this: https://tenstorrent.slack.com/archives/C07A30HRX7C/p1730460299618899 So everyone working on frontends can have their own GH tasks about ops bringup, but you also have to open separate task under the `MLIR Dialect / Passes` component in case that you have to make changes in tt-mlir TTIR/TTNN dialect/passes for that specific op. That way we won't have clashes in op bringup in TTIR/TTNN.

File: lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp
Comment (mrakitaTT) [reply lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp:698]: @sdjordjevicTT Thank you, I will share this with the team then so we are all in sync. If you want I can make a page in the docs similar to [this one](https://docs.tenstorrent.com/tt-mlir/adding-an-op.html) or maybe even under that one, and then you guys can add to it as you continue working on design improvements for the passes.

Regarding having one or multiple TTIR ops, we've discussed offline and agreed to go with a single op in TTIR.

Comment (mrakitaTT) [reply lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp:698]: Resolving this conversation now to unblock @uazizTT, since this PR is in compliance with the guidelines above.

File: include/ttmlir/Dialect/TTIR/Pipelines/TTIRPipelines.h
Comment (mrakitaTT) [inline include/ttmlir/Dialect/TTIR/Pipelines/TTIRPipelines.h:1]: Are the changes in this file result of some weird merge with #926?

File: runtime/lib/ttnn/operations/eltwise/tertiary/where.cpp
Comment (mrakitaTT) [inline runtime/lib/ttnn/operations/eltwise/tertiary/where.cpp:13]: Seems like you forgot to remove this line, not sure how is it compiling in CI O.o

File: test/ttmlir/Dialect/TTNN/simple_where.mlir
Comment (mrakitaTT) [inline test/ttmlir/Dialect/TTNN/simple_where.mlir:6]: Can you please also add a CHECK for this line, to make sure the whole combination is being lowered properly

**Review by mrakitaTT — CHANGES_REQUESTED** [review]

File: runtime/lib/ttnn/operations/eltwise/tertiary/where.cpp
Comment (jnie-TT) [inline runtime/lib/ttnn/operations/eltwise/tertiary/where.cpp:19]: What are we doing here? Aren't op->on_true(), op->on_false() tensor ref pointers? 
```
table WhereOp {
  pred: tt.target.TensorRef;
  on_true: tt.target.TensorRef;
  on_false: tt.target.TensorRef;
  output: tt.target.TensorRef;
}
```
Also I think the correct way to implement this is to create an eltwise ternary file and add this op here, since ttnn models this op as an eltwise ternary op.

**Review by jnie-TT — CHANGES_REQUESTED** [review]

File: runtime/lib/ttnn/operations/eltwise/tertiary/where.cpp
Comment (jnie-TT) [reply runtime/lib/ttnn/operations/eltwise/tertiary/where.cpp:19]: Also please add a silicon test for this under Silicon/TTNN

Comment (mrakitaTT) [reply runtime/lib/ttnn/operations/eltwise/tertiary/where.cpp]: > The reason this doesn't break anything is probably because this file wasn't added to runtime/lib/ttnn/operations/CMakeLists.txt

Oh, that explains [this](https://github.com/tenstorrent/tt-mlir/pull/852#discussion_r1830074361) :) 

> Also I think the correct way to implement this is to create an eltwise ternary file and add this op there, since ttnn models this op as an eltwise ternary op.

@jnie-TT Can you check [this](https://github.com/tenstorrent/tt-mlir/pull/852#discussion_r1819112960) comment please? In short, Metal currently also has a separate file for where op since it has different parameters from other ternary ops.

Comment (jnie-TT) [reply runtime/lib/ttnn/operations/eltwise/tertiary/where.cpp]: @mrakitaTT Yeah we should create a new file, and we should encapsulate the op in the ternary namespace, possibly also modelling this op as ternary in the flatbuffer schema.

Comment (uazizTT) [reply runtime/lib/ttnn/operations/eltwise/tertiary/where.cpp]: > @mrakitaTT Yeah we should create a new file, and we should encapsulate the op in the ternary namespace, possibly also modelling this op as ternary in the flatbuffer schema and maybe in the compiler as well.

I think that unless we can find two or more tertiary ops that behave the same way like we had for binary and unary ops, there is no point of creating a common infrastructure for them and we should keep them separate like I did with where op and be treated like a standalone op.

Comment (mrakitaTT) [reply runtime/lib/ttnn/operations/eltwise/tertiary/where.cpp]: > Yeah we should create a new file, and we should encapsulate the op in the ternary namespace

Yeah, Usman moved it into separate file and made a ternary folder, but you are right he misnamed it into `tertiary` and didn't update the namespace, good catch on those, @uazizTT can you please fix?

> possibly also modelling this op as ternary in the flatbuffer schema and maybe in the compiler as well

@jnie-TT As you can see in that comment I didn't suggest Usman to make other changes yet because Metal currently keeps `where` op in a separate file and there is no connection with other ternary ops except it being in the same folder as them. It also has a different interface than other ternary ops, so I don't see much that we can do in the compiler yet. I think the best we can currently do is to match Metal structure and keep where op separate from other ternary ops, and when they make a change we can make a change too. But I'll let you and Kyle decide of course :)

Comment (jnie-TT) [reply runtime/lib/ttnn/operations/eltwise/tertiary/where.cpp]: Where op should be a standalone file in runtime because it's a standalone file in ttnn. 
I think the goal is to model as closely to ttnn as possible, so we should model this closely to ttnn from the start instead of having to merge the ops later. Even if we only have 1 ternary op we should model it as a ternary op instead of a standalone op. We did the same for reduce ops, where initially we had only reduce sum, but we still modelled it in the bracket of reduce op.

Comment (jnie-TT) [reply runtime/lib/ttnn/operations/eltwise/tertiary/where.cpp]: @mrakitaTT I agree that we should make it a separate file than other ternary ops. But I think that in at least in the ttnn program flatbuffer schema we should model it as a ternary op. For example unary.hpp and unary_composite.hpp are separate files, but we still model them as eltwise unary ops in flatbuffer. If we need extra parameters we can add an interface for that. Looking at the interface, it's not that much different from other ternary ops:
```
Tensor WhereOperation::invoke(
    uint8_t queue_id,
    const Tensor& predicate,
    const Tensor& value_true,
    const Tensor& value_false,
    const std::optional<MemoryConfig>& output_mem_config,
    std::optional<Tensor> output_tensor) {

    return ternary_utils::where_impl(queue_id, predicate, value_true, value_false, output_mem_config.value_or(predicate.memory_config()), output_tensor);
}
```
Other ternary ops follow roughly the same structure, while some may require an extra float argument.

Comment (mrakitaTT) [reply runtime/lib/ttnn/operations/eltwise/tertiary/where.cpp]: > I think that unless we can find two or more tertiary ops that behave the same way like we had for binary and unary ops, there is no point of creating a common infrastructure for them and we should keep them separate like I did with where op and be treated like a standalone op.

@uazizTT I agree about treating where op separately for now, but as for the common infrastructure part I think that as soon as we have the first ternary op we should make a support for it in a general way so that other ternary ops can be added in the same file. Same as what @jnie-TT said in the comment above, we are overlapping with comments right now :)

Comment (jnie-TT) [reply runtime/lib/ttnn/operations/eltwise/tertiary/where.cpp]: If it's too much work or it's unclear how this op will interface with other ternary ops in some common infra, then yeah we can add it standalone for now, but please leave a note that this should be considered when creating an infra for ternary.

Comment (mrakitaTT) [reply runtime/lib/ttnn/operations/eltwise/tertiary/where.cpp]: > I agree that we should make it a separate file than other ternary ops. But I think that at least in the ttnn program flatbuffer schema we should model it as a ternary op. For example unary.hpp and unary_composite.hpp are separate files, but we still model them as eltwise unary ops in flatbuffer. If we need extra parameters we can add an interface for that.

@jnie-TT Sure, makes sense

Comment (jnie-TT) [reply runtime/lib/ttnn/operations/eltwise/tertiary/where.cpp]: Thanks guys for the discussion! I'm trying to make runtime as close to TTNN as possible and limiting the amount of workarounds when needed. I know it can be cumbersome at times but as we bring up more ops it'd be desirable if we can keep it organized in this manner.

Comment (mrakitaTT) [reply runtime/lib/ttnn/operations/eltwise/tertiary/where.cpp]: @jnie-TT np, so just to make clear, is this the set of required changes that we converged to?
- Rename folder and include guard to use `ternary` instead of `tertiary`
- Fix the namespace
- Create common ternary ops interface in flat buffer and then extend interface for where op from it

File: lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp
Comment (svuckovicTT) [inline lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp:831]: No need to template it, if it's going to be used for one op, as the name suggets.

Comment (svuckovicTT) [inline lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp:831]: Pattern is called `SelectOpConversionPattern` but it converts `ttir::WhereOp` to `ttnn::WhereOp`.

**Review by svuckovicTT — APPROVED** [review]

File: include/ttmlir/Dialect/TTNN/IR/TTNNOps.td
Comment (jnie-TT) [inline include/ttmlir/Dialect/TTNN/IR/TTNNOps.td:153]: Nit: ternary

**Review by jnie-TT — APPROVED** [review]
This is awesome. I forgot that in the flatbuffer the inputs is actually an array of variable-number elements, so we can integrate this into eltwiseop directly instead of creating a ternary op type, even better than the original plan! Thanks for looking at this!

**Review by tapspatel — APPROVED** [review]
looks good! only change is can you add a test under Silicon/TTNN/perf_unit for where op? otherwise looks good

**Pull request merged**
